### ISSUE TITLE
New system attribute `encounterable fleet`, that represent encountered fleets rather than spawned fleets.

### ISSUE THREAD

**Issue by Pshy0** [endless-sky/endless-sky#9411]
### Problem Description

To me, the fleets defined in systems should represent what you can encounter by being in that system, not because fleets magically spawn in by the presence of the player, but because they may come from a planet or a connected system.

With the current implementation, it feels like the entire universe has no ship activity until you're here. You do not find any fleet, you summon them by your presence. Nothing is happening until you're here. Fleets only arrive or take off because of the presence of the player, but otherwise does not exist. And also that ships travel randomly with no plan.

If I define a trade route in a plugin, lets say for specialized huge container ships, I would need to set each system on that route capable of spawning the specific ships that use that trade route. At this point it's intuitive, but then I don't expect the player to meet any of those ships outside of that route just because the player entered and then left it, because I never added those fleets to any other nearby system. With the current implementation, I'd have to prohibit those fleets from entering the systems surrounding that route using the `restricted` attribute (without mentioning I'd currently need a separate government just for that fleet), and any update to the route will require updating not only the route's systems but also connected systems. And I don't mention plugins that may connect system to others that you did not account for (with links or just being close enough to jump). Doing it this way feels not right, hacky, and more complicated than it should be.

Aside from what ship is encountered in what system, this can also be an issue with the frequency of ships, which choose a system randomly to jump to, even if one has an higher chance of seeing a fleet than another. Which could cause a system to temporarily be perceived as having an higher spawn rate than it should.

### Related Issue Links

- #6489 add a way to simulate this, but only per-government, and it requires maintaining all systems around the ones you want your fleets to be encounterable from. Which is explained by.
- #6864 suggests something similar but more complex.
- This would help fixing #8876, #8986, #9298, #9299.
- If implemented with the weights, this could help fixing #4039.

### Desired Solution

Add an `encounterable fleet` attribute, that represent fleets you might encounter in a system, rather than fleets that your presence in that system would cause to spawn.

The difference is that fleets spawned with this attribute would only travel to systems where this fleet might be encountered. It would be more maintainable that blocking fleets in nearby system's definitions.

Perhaps this can be made even better by, instead of the attribute's second argument being a spawn delay, to be a spawn per minute (accepting float numbers). This way, the `add` attribute would not require an exception to work as expected. The syntax would be:
```
`encounterable fleet` <fleet_name> <count_per_minute>
```
or
```
`encounterable fleet`
    <fleet_name> <count_per_minute>
    ...
```

Additionally, count_per_minute could serve as a weight to determine to which system a ship will jump, rather that it being random with equal weights.

### Alternative Approaches

- Changing the current `fleet` behavior itself, as this was originally suggested in #8876.
- #6489 add a way to simulate this, but only per-government, and it requires maintaining all systems around the ones you want your fleets to be encounterable from.
- #6864 suggests something similar but more complex.
- Define the path that a fleet has planed to take when it spawns, rather than every jump (but using the same conditions).

**Comment by Saugia** [comment]
See [MZ's comment on an older issue.](https://github.com/endless-sky/endless-sky/issues/2446#issuecomment-298791430) Adding this kind of capability could be quite taxing on system load and adds more complexity to the game that is necessary for not as much gain in comparison.

**Comment by Pshy0** [comment]
> See https://github.com/endless-sky/endless-sky/issues/2446#issuecomment-298791430 Adding this kind of capability could be quite taxing on system load and adds more complexity to the game that is not necessary, for only so much gain in comparison.

I read it, but the linked comment was about simulating what is happening in other systems. This issue is not about simulating what is happening in other systems at all and you do not need to do that to implement it. All that should be needed to implement this one is to, instead of choosing from connected systems to which one a ship jumps, choose from connected systems that also have this fleet defined in. It should not be more expensive than the `restricted` attribute. And this fleet would not be encountered again in a system it's not defined in. And it's not happening every frame.

> System fleets don't need to be so precise that they simulate proper encountered fleets, and having persistent fleets defined at any given time, as opposed to being based around the player, especially if it's for everywhere in the game, is a lot of extra work and data the game would have to keep track of.

This issue is not either about persistent fleets. There is no more or less fleets stored by the game or more data to be defined.
In fact, if this replaces `restricted`, there will be **less** data to define, for better accuracy.

> These sorts of intentionally defined fleets should be for specific cases or inconsistencies

Why? Is there any case where it is useful that a fleet ends up in a system you would not normally see it, because it does not spawn there?

**Comment by Pshy0** [comment]
I updated the description to make it clearer.

**Comment by Saugia** [comment]
I wasn't sure what you were suggesting, as...

> The difference is that fleets spawned with this attribute would only travel to systems where this fleet might be encountered. It would be more maintainable that blocking fleets in nearby system's definitions.

...reads "Fleets that are spawned are traveling through systems only where they are found", aka, this is happening in the background. So *is* this based on fleets *after* they are spawned because of player presence? If so, I'm not sure "To me, the fleets defined in systems should represent what you can encounter by being in that system, not because fleets magically spawn in by the presence of the player..." reads properly to that, and that's what I was gathering from this issue.

Are you suggesting to further nuance fleets here beyond `restricted`, or are you suggesting this as an alternative to `restricted`?

> Why? Is there any case where it is useful that a fleet ends up in a system you would not normally see it, because it does not spawn there?

This is a matter of simplicity. Why do we need to nuance fleets so much that it's based on encounterability? The main concerns (Wanderers, FW, Pug) are in discussion and are specific examples. As Amazinite was putting it, there's not really a reason to be blanketing this sort of thing across every spawned fleet. 

If main concerns are addressed, what benefit do we gain from doing it everywhere?  What is wrong with fleets exiting systems or entering systems after being spawned? Why should AI fleets be limited by this capacity? Is there a balance reason? If so, we can again use `restrictions` for that. Is it because it doesn't make sense? Ditto. This seems like extra complexity for the sake of something that, as you state, "makes sense to you".

**Comment by Saugia** [comment]
> One thing that this implementation is making worse, is the feeling that the entire universe has no ship activity until you're here. You do not find any fleet, you summon them by your presence. Fleets only arrive or take off because of the presence of the player, but otherwise does not exist. And also that ships travel randomly with no plan.

Also, I'm not sure where you're getting this from.

If I take off from a planet, in my given case, there is a fleet already in the system. However, this does not mean that every planet you take off from a fleet will be spawned immediately. Sometimes they jump in after, or take off after you.

If I hyperspace from one system to another, in my case where I tested, there is already a fleet waiting in the system. This wasn't spawned just because I jumped into the system, it was already there. Again, this isn't always going to be the case, and sometimes things will enter later or sooner, but a fleet was already there.

In my case, when I first jump drived to another system, there was no fleet, however, some did enter shortly after. I did it again, second time, there was a fleet in the system I went to already waiting. Note I tested all of these with human systems, but it isn't any different for alien races. Jump to a Korath Exiles system - there will likely be ships waiting.

Regardless of how the system spawns the fleets, what you're saying here is not true. Fleets can be in the system upon arrival, or they can enter at a different time. I don't see how the suggestions here change that in a beneficial way (or beyond what it already does).

**Comment by Pshy0** [comment]
> Fleets that are spawned are traveling through systems only where they are found

This is the desired behavior. Or at least, this should be the perception you must have as the player. 

> aka, this is happening in the background

Not more than it already is.

> So is this based on fleets after they are spawned because of player presence?

Yes

> Are you suggesting to further nuance fleets here beyond `restricted`, or are you suggesting this as an alternative to `restricted`?

It's an alternative that would work immediately for any defined fleet rather than having to think of what system to restrict when making a fleet or what fleets to restrict when making a system. I see `restricted` as a blacklist you have to think about every-time you do a change, being in the data or events, and this proposal as a whitelist you do not have to maintain at all.

> This is a matter of simplicity.

I understand that. It is simpler to spawn ships and let them travel anywhere rather than controlling where they are traveling to. Sometime you initially implement something in a simpler way, so that it just works, even if there is inconsistencies. This is what **I believe** the `fleet` attribute is: _The need for fleets to be encountered was solved by this `fleet` attribute that represent what can be encountered in a system. But it was implemented the "simpler" way that just spawns ship and does not care about where they jump. As long as you stay in a system, it works as intended. But when you jump, you can find fleets in systems you would not have found them otherwise._ This can be improved.

> The main concerns (Wanderers, FW, Pug) are in discussion and are specific examples.

I do not think they are examples, I think they are the general rule. For absolutely all fleets, there is locations you do not want them to be encountered in. Korath other than raids can travel to human space, Remnants can travel to human space too, Quargs can jump to Coalition space, Merchants can get themselves in Unfettered space... It may happen rarely enough to not be a concern, but this is what the current implementation allows. And when a plugin adds systems in, this create even more cases.

> If main concerns are addressed, what benefit do we gain from doing it everywhere?

More maintainability. Updating what fleets can be encountered in a system would only involve updating that system, instead of all systems in jump range. And this would fix all the cases without having to think about each one individually.

> What is wrong with fleets exiting systems or entering systems after being spawned?

This issue is not about preventing ships from jumping.

> Why should AI fleets be limited by this capacity compared to the player?

The player is not a random NPC. He is not one of those merchants with boring traditional routes or a navy member bound to protecting a specific space.

If fleets were supposed to have the same freedom as the player, they should be spawned in absolutely every system (at least with a small frequency), because they would then have a chance to end there. Add Militia ships to syndicate space, more Korath in human space, and deeper into human space, and remnants to human space...

> This seems like extra complexity for the sake of something that, as you state, "makes sense to you".

It is not more complex, it makes maintaining fleets, systems, and events altering them, simpler. This suggestion would fix all the mentioned cases, and similar cases, current and future. Without it, you'll have to either worry about this when modifying fleets, systems and events for eternity, or wait for someone to make an issue for every case they encounter, knowing that there will always be one that will never be fixed.

**Comment by Pshy0** [comment]
> they're leaving the system they came from and were originally spawned there.

Not it's not the way it currently works, and the Pug example demonstrates that.

**Comment by Saugia** [comment]
>>Fleets that are spawned are traveling through systems only where they are found

>This is the desired behavior. Or at least, this should be the perception you must have as the player.

What I'm getting from this is:

If I have a fleet, say:
```
fleet "Large Fleet"
government "Republic"
      variant Freighter
      variant Shuttle
      variant Sparrow
```
That, this fleet should only spawn in systems it makes sense to spawn them in? If so, then that means human systems. What's the problem here then?

---
>>So is this based on fleets after they are spawned because of player presence?

>Yes

If so, then

>>aka, this is happening in the background

> Not more than it already is."

makes no sense because there is no way to change how a fleet spawns outside of the player's presence without having to define more fleets or instances of them in the background, which is still tied to the game loading, aka landing, taking off, and map position, aka the player's presence.

---
>It's an alternative that would work immediately for any defined fleet rather than having to think of what system to restrict when making a fleet or what fleets to restrict when making a system. I see restricted as a blacklist you have to think about every-time you do a change, being in the data or events, and this proposal as a whitelist you do not have to maintain at all.

Whether or not something is "blacklisted" is not something the player is going to think. This is coming from a development perspective. The player isn't going to see or think this while playing the game. At most, it would be, "Where are the Pug ships jumping to?" or "Why aren't the Wanderer ships jumping to Hai space"? Which both have in-lore reasons to be limited that way. The player isn't going to think, "oh they're just blocked by a tool in the game".

Having to make a change in order for something to work properly is normal. Making an overall mechanical change to the game in order to prevent such a need is completely out of scope here.

---
> I understand that. It is simpler to spawn ships and let them travel anywhere rather than controlling where they are traveling to. Sometime you initially implement something in a simpler way, so that it just works, even if there is inconsistencies. This is what I believe the fleet attribute is: The need for fleets to be encountered was solved by this fleet attribute that represent what can be encountered in a system. But it was implemented the "simpler" way that just spawns ship and does not care about where they jump. As long as you stay in a system, it works as intended. But when you jump, you can find fleets in systems you would not have found them otherwise. This can be improved.

Yes, it can be improved, which again is what the `restricted` attribute is for. Even if the tool is to prevent inconsistencies, it is not simply a blocking tool is not working efficiently. It does what it's supposed to do, and there's no reason to overhaul the system entirely to include it, when these individual cases *are* the problems, and nothing else is.

---
> I do not think they are examples, I think they are the general rule. For absolutely all fleets, there is locations you do not want them to be encountered in. Korath other than raids can travel to human space, Remnants can travel to human space too, Quargs can jump to Coalition space, Merchants can get themselves in Unfettered space... It may happen rarely enough to not be a concern, but this is what the current implementation allows. And when a plugin adds systems in, this create even more cases.

Then we simply add these to the restrictions. These cases are specifically that, *restricting fleets so that they don't go beyond a boundary that makes sense. Converting that to "an overall change to the fleet system to include these cases" is not the same thing. This is not a matter of "encounterabilty", it's a matter of *restriction*. Yes, there is a general rule that fleets should not be going to places that don't make sense. Restricting them from doing so is all that needs to be done. If these are the main points, then having restriction is all that is necessary to fix them, not add a whole new way or tag on how fleets work that changes the system.

Also, this is not something we should necessarily be deciding overall anyway. It might make sense in some content areas to have this overlap, and others not. Which again, restriction does just fine.

---
>>If main concerns are addressed, what benefit do we gain from doing it everywhere?

>More maintainability. Updating what fleets can be encountered in a system would only involve updating that system, instead of all systems in jump range. And this would fix all the cases without having to think about each one individually.

Once again, you're adding overhead for something that is only required for certain cases. Even if there are more than a few cases. Restriction is all that's needed, not an overall change. Also, updating "what fleets can be encountered in a system only involving that system" is around the same work as "all systems in jump range". The work needed between this is negligible. We are thinking about these cases individually, because they *are individual cases*, not a map-wide problem, aka, not every single system or territory.

---
>>Why should AI fleets be limited by this capacity compared to the player?

>The player is not a random NPC. He is not one of those merchants with boring traditional routes or a navy member bound to protecting a specific space.

If fleets were supposed to have the same freedom as the player, they should be spawned in absolutely every system (at least with a small frequency), because they would then have a chance to end there. Add Militia ships to syndicate space, more Korath in human space, and deeper into human space, and remnants to human space...

This doesn't answer my question. Why limit every case where a ship can jump or fly out a direction from their original spawn territory? And, if it does, and it doesn't make sense, again, it's case-specific, not the entire map. 

Fleets *do* "spawn" in every system, it just takes the instancing and player's presence, yes. But it's not on the direct presence of the player, it occurs in nearby systems to the player too. Having it work through this instancing *already is* that level of freedom, within acceptable CPU bounds. Once again, if a fleet doesn't make sense where it's going, restrict that case. It's that simple. It's not inefficient. It's all that's necessary. Adding an overarching method to encompass inconsistencies is not efficient in itself. If it's to add more capability or variety, perhaps, but in what's presented here, it's to change how fleets work overall.

>It is not more complex, it makes maintaining fleets, systems, and events altering them, simpler. This suggestion would fix all the mentioned cases, and similar cases, current and future. Without it, you'll have to either worry about this when modifying fleets, systems and events for eternity, or wait for someone to make an issue for every case they encounter, knowing that there will always be one that will never be fixed.
 
It is adding something that is otherwise done using the restriction method. It is adding something that we don't need. The restriction method also fixes all current cases, and future ones. With this suggestion, *it's not any different, you will still have to worry about adding these tags to future cases for "all of eternity". Hence, it is adding more than what is necessary. If we add the restriction method, we *are* addressing all current cases with it, and anything that comes up in future content will have it in the game to help prevent it. It's not like we're going to add restriction, and then after that, add more fleets that don't make sense and wait for someone to open an issue before we fix it. 

Once we add restriction, we will keep it in mind when adding more fleets with this capability. I think you're way overemphasizing the "restriction" and "block" denotation of the method as negative. It's simply a method that can resolve all current cases and can be used for any that are introduced in the future, which is the same thing you're suggesting here, except without a need to blanket everything and fleets overall, which is more work in itself. 

If you have a big country farm, and one of the gates broke, you fix that gate. You don't buy or construct a whole new gate system because one part of it broke. If several gates are broke, and the method to fix the first one fixes those fine too, then there's still no reason to invest in a whole new system.

**Comment by Saugia** [comment]
To avoid going in a circle with a bunch of replies, I'll leave my comments to this.

Adding an `encounterable fleet` attribute for systems is just as much work as `restricting` system fleets, if not more. You're looking at changing every system, especially with nuances such as count and time, as opposed to the specific inconsistent fleets.

`Restriction` is all that is necessary. If it solves the problem, there's no reason to add an overall change to how system fleets work. If the concern is future examples, `restriction` can address those too.

Those are my thoughts on the matter.

**Comment by Pshy0** [comment]
I feel like you believe my suggestion would be to, not only add an attribute, change all existing attributes, convert delays to frequencies, and change the current `fleet` behavior. That would indeed be a lot of work. But this is **NOT** a **SINGLE** suggestion, those are separate suggestions to fix this issue. The preferred one being to change the current `fleet` behavior, the alternative being to add `encounterable fleet` so that, at least, plugin creators have an option to make their data more maintainable, if you do not want the game data to be.

You say `restricted` is sufficient in all cases. I explained that `restricted` could fix some of the worries but not all. I explained why it's also much less maintainable. It's like fixing a breach in a paper wall with ducktape, not to have to replace the paper wall by a stronger wood one. 

Regarding "spawning", I gave my definition of it. It looks like you are interpreting my words with a different definition that says a fleet can be spawned, but also not yet being instantiated / existing / able to move around. And It appears we disagree on how ships are spawning, and I do not know a better example than the Pug one i gave to explain this. From my point of view, this suggestion would not and does not need to change spawning at all.

I explained the new attribute further and named this issue accordingly, because [Amazinite recently expressed that he did not want to prevent the `fleet`s from wandering around](https://github.com/endless-sky/endless-sky/issues/8876#issuecomment-1745925151), and I did not want this proposal to be misinterpreted as only being about one specific vanilla content, leading to the renaming of the issue into something specific to only one example, ignoring the real problem - again. But I do disagree with the current `fleet` behavior, which is responsible for the need for `restricted`, and is responsible for issues planed to be resolved by it.

**Comment by Saugia** [comment]
The way the issue is formatted, adding an `encounterable fleet` is the primary goal here, and a change in fleet behavior overall is the alternative. I'm stating that the former is an unnecessary approach.

> ...so that, at least, plugin creators have an option to make their data more maintainable, if you do not want the game data to be.

We do not add code or mechanics to the game solely for plug-ins. When it comes to adding new capabilities to the game, we take them for their value in vanilla content. If the feature is not desired or fits in vanilla content, we are not going to add it just for plug-ins.

> You say restricted is sufficient in all cases. I explained that restricted could fix some of the worries but not all. I explained why it's also much less maintainable. It's like fixing a breach in a paper wall with ducktape, not to have to replace the paper wall by a stronger wood one.

This is an oxymoron. The issue `restriction` is resolving is with these specific cases. There is no overall case, or overall "wall", that needs replacing. Now, if you're suggesting to change fleets to incorporate representations such as trade routes displayed by ship spawns, that's one thing - but this is where it comes to being unnecessary complexity for the game's environment. We don't need to nuance things such as trade routes based on AI ship representation.

I don't think this is a matter of a problem of how fleets work. I think this is a desire to expand on how they work. In terms of problems, `restriction` solves our cases. In terms of edge cases or cases you'd like to add, that's more expandability. That's probably where the confusion is arising.

**Comment by ziproot** [comment]
An alternative approach not mentioned here is to have an attribute that can be defined for fleets or NPCs, which specifies a [location filter](https://github.com/endless-sky/endless-sky/wiki/LocationFilters). The fleet or NPC would only be able to jump to a system in that location filter. This allows for far more flexibility than just government restrictions, so it could, for instance, allow Republic ships to jump to Free Worlds systems, but only if they are within N jumps. It could also allow for creating Pug pocket dimensions that NPCs cannot travel to, while still allowing them to travel to Deneb, by simply blocking NPCs from traveling to systems with the `inaccessible` attribute. Currently, this would also require giving the pocket dimension systems a dummy government and setting a travel restriction for that government. Thoughts?
